File: cloud-run/modules/drive_client.py
"""
Google Drive APIクライアント
ファイルの取得、移動、リネームなどの操作
"""
import time
import logging
import traceback
from typing import Optional, List, Dict, Any
from googleapiclient.discovery import build
from googleapiclient.http import MediaIoBaseDownload
from google.oauth2 import service_account
import io

logger = logging.getLogger(__name__)


class DriveClient:
    """Google Drive APIクライアント"""
    
    def __init__(self, credentials=None):
        """
        初期化
        
        Args:
            credentials: 認証情報（Noneの場合はデフォルト認証を使用）
        """
        if credentials:
            self.service = build('drive', 'v3', credentials=credentials)
        else:
            # デフォルト認証（Cloud Run環境ではサービスアカウント）
            self.service = build('drive', 'v3')
    
    def get_file(self, file_id: str) -> Optional[dict]:
        """
        ファイル情報を取得
        
        Args:
            file_id: ファイルID
            
        Returns:
            ファイル情報
        """
        try:
            file = self.service.files().get(
                fileId=file_id,
                fields='id, name, mimeType, parents'
            ).execute()
            return file
        except Exception as e:
            logger.error(f"ファイル情報取得エラー: {e}")
            return None
    
    def download_file(self, file_id: str) -> Optional[bytes]:
        """
        ファイルをダウンロード（堅牢化版）
        """
        import time
        import traceback
        import socket
        
        # デフォルトソケットタイムアウトを120秒に設定
        socket.setdefaulttimeout(120)
        
        max_retries = 5
        for attempt in range(max_retries):
            try:
                logger.info(f"ダウンロード開始（試行 {attempt+1}/{max_retries}）: {file_id}")
                
                # サービスオブジェクトの状態が悪い可能性があるので、リトライ時は再構築を検討
                if attempt > 0:
                    logger.info("サービスオブジェクトを再構築してリトライします")
                    self.service = build('drive', 'v3')

                request = self.service.files().get_media(fileId=file_id)
                file_content = request.execute()
                
                logger.info(f"ダウンロード完了: {len(file_content)} bytes")
                return file_content
            
            except Exception as e:
                # 最終試行で失敗
                if attempt == max_retries - 1:
                    logger.error(f"ダウンロード最終失敗: {e}")
                    logger.error(f"Traceback: {traceback.format_exc()}")
                    return None
                
                logger.info(f"ダウンロード失敗（リトライ待ち）: {e}")
                wait_time = 2 ** attempt
                time.sleep(wait_time)
        
        return None
    
    def move_file(
        self, 
        file_id: str, 
        new_parent_id: str,
        current_parent_id: Optional[str] = None
    ) -> bool:
        """
        ファイルを移動
        
        Args:
            file_id: ファイルID
            new_parent_id: 移動先フォルダID
            current_parent_id: 現在の親フォルダID（Noneの場合は自動取得）
            
        Returns:
            成功したかどうか
        """
        try:
            # 現在の親フォルダIDを取得
            if not current_parent_id:
                file = self.get_file(file_id)
                if not file or 'parents' not in file:
                    logger.error("親フォルダIDが取得できません")
                    return False
                current_parent_id = file['parents'][0]
            
            # ファイルを移動
            self.service.files().update(
                fileId=file_id,
                addParents=new_parent_id,
                removeParents=current_parent_id,
                fields='id, parents'
            ).execute()
            
            logger.info(f"ファイル移動成功: {file_id}")
            return True
        except Exception as e:
            logger.error(f"ファイル移動エラー: {e}")
            return False
    
    def rename_file(self, file_id: str, new_name: str) -> bool:
        """
        ファイル名を変更
        
        Args:
            file_id: ファイルID
            new_name: 新しいファイル名
            
        Returns:
            成功したかどうか
        """
        try:
            self.service.files().update(
                fileId=file_id,
                body={'name': new_name}
            ).execute()
            
            logger.info(f"ファイル名変更成功: {new_name}")
            return True
        except Exception as e:
            logger.error(f"ファイル名変更エラー: {e}")
            return False
    
    def create_folder(
        self, 
        folder_name: str, 
        parent_id: str
    ) -> Optional[str]:
        """
        フォルダを作成
        
        Args:
            folder_name: フォルダ名
            parent_id: 親フォルダID
            
        Returns:
            作成されたフォルダID
        """
        try:
            file_metadata = {
                'name': folder_name,
                'mimeType': 'application/vnd.google-apps.folder',
                'parents': [parent_id]
            }
            
            folder = self.service.files().create(
                body=file_metadata,
                fields='id',
                supportsAllDrives=True
            ).execute()
            
            logger.info(f"フォルダ作成成功: {folder_name}")
            return folder.get('id')
        except Exception as e:
            logger.error(f"フォルダ作成エラー: {e}")
            return None
    
    def find_folder(
        self, 
        folder_name: str, 
        parent_id: str
    ) -> Optional[str]:
        """
        フォルダを検索
        
        Args:
            folder_name: フォルダ名
            parent_id: 親フォルダID
            
        Returns:
            フォルダID（見つからない場合はNone）
        """
        try:
            query = f"name='{folder_name}' and '{parent_id}' in parents and mimeType='application/vnd.google-apps.folder' and trashed=false"
            results = self.service.files().list(
                q=query,
                fields='files(id, name)',
                supportsAllDrives=True,
                includeItemsFromAllDrives=True
            ).execute()
            
            files = results.get('files', [])
            if files:
                logger.info(f"既存フォルダ発見: {folder_name} ({files[0]['id']})")
                return files[0]['id']
            logger.info(f"フォルダが見つかりません: {folder_name} in {parent_id}")
            return None
        except Exception as e:
            logger.error(f"フォルダ検索エラー: {e}")
            return None
    
    def get_or_create_folder(
        self, 
        folder_name: str, 
        parent_id: str
    ) -> Optional[str]:
        """
        フォルダを取得または作成
        
        Args:
            folder_name: フォルダ名
            parent_id: 親フォルダID
            
        Returns:
            フォルダID
        """
        folder_id = self.find_folder(folder_name, parent_id)
        if folder_id:
            return folder_id
        return self.create_folder(folder_name, parent_id)

    def list_files(self, query: str, limit: int = 100) -> List[Dict]:
        """クエリに基づいてファイルを検索"""
        try:
            results = self.service.files().list(
                q=query,
                pageSize=limit,
                fields="nextPageToken, files(id, name, mimeType, size, owners, quotaBytesUsed)"
            ).execute()
            return results.get('files', [])
        except Exception as e:
            logger.error(f"ファイル検索エラー: {e}")
            return []

    def list_files_in_folder(self, folder_id: str, limit: int = 100) -> List[Dict]:
        """フォルダ内のファイルを検索"""
        query = f"'{folder_id}' in parents and trashed=false"
        return self.list_files(query, limit)

    def delete_file(self, file_id: str) -> bool:
        """ファイルを削除"""
        try:
            self.service.files().delete(fileId=file_id).execute()
            logger.info(f"ファイル削除成功: {file_id}")
            return True
        except Exception as e:
            logger.error(f"ファイル削除エラー: {e}")
            return False

    def cleanup_service_account_storage(self) -> Dict[str, Any]:
        """SAのストレージをクリーンアップ（ゴミ箱を空に & 所有ファイルを削除）"""
        stats = {'deleted_files': 0, 'freed_bytes': 0, 'errors': []}
        try:
            # 1. ゴミ箱を空にする
            try:
                self.service.files().emptyTrash().execute()
                logger.info("ゴミ箱を空にしました")
            except Exception as e:
                logger.warning(f"ゴミ箱のクリーンアップ失敗: {e}")
            
            # 2. SAが所有するファイルを検索 ('me' in owners)
            # NotebookLM同期ファイルなどの重要なものは除外したいが、
            # SAが所有している＝ユーザーに転送できていない＝容量圧迫の原因。
            # 今回は緊急対応として、SA所有の全ファイルを対象とする（ユーザー同意済みと仮定）
            # ただし、フォルダは消さないようにmimeTypeでフィルタしてもよいが、
            # deleteは再帰的ではないのでフォルダなら中身だけ消える？いや、フォルダも消える。
            # 安全のため、application/vnd.google-apps.folderは除外する。
            query = "'me' in owners and trashed = false and mimeType != 'application/vnd.google-apps.folder'"
            files = self.list_files(query, limit=100)
            
            for file in files:
                file_id = file['id']
                file_name = file.get('name', 'Unknown')
                file_size = int(file.get('quotaBytesUsed', 0))
                
                logger.info(f"Deleting SA-owned file: {file_name} ({file_size} bytes)")
                
                if self.delete_file(file_id):
                    stats['deleted_files'] += 1
                    stats['freed_bytes'] += file_size
                else:
                    stats['errors'].append(f"Failed to delete {file_name}")
                    
            return stats
        except Exception as e:
            logger.error(f"クリーンアップエラー: {e}")
            stats['errors'].append(str(e))
            return stats

File: cloud-run-go/internal/service/drive_client.go
package service

import (
	"context"
	"fmt"
	"io"
	"log"
	"sync"
	"time"

	"github.com/leo-sagawa/homedocmanager/internal/model"
	"golang.org/x/oauth2"
	"google.golang.org/api/docs/v1"
	"google.golang.org/api/drive/v3"
	"google.golang.org/api/option"
)

// DriveClient はGoogle Drive APIクライアント
type DriveClient struct {
	service     *drive.Service
	docsService *docs.Service
	oauthCreds  *OAuthCredentials
	folderCache map[string]string // key: "parentID:folderName", value: folderID
	folderMu    sync.Mutex
}

// NewDriveClient は新しいDriveClientを作成
func NewDriveClient(ctx context.Context) (*DriveClient, error) {
	// OAuth認証情報を取得
	creds, err := GetOAuthCredentials(ctx)
	if err != nil {
		log.Printf("Warning: Failed to get OAuth credentials for Drive, falling back to Service Account: %v", err)
		// サービスアカウント認証にフォールバック
		service, err := drive.NewService(ctx, option.WithScopes(drive.DriveScope))
		if err != nil {
			return nil, fmt.Errorf("failed to create drive service: %w", err)
		}
		docsService, err := docs.NewService(ctx, option.WithScopes(docs.DocumentsScope))
		if err != nil {
			return nil, fmt.Errorf("failed to create docs service: %w", err)
		}
		return &DriveClient{
			service:     service,
			docsService: docsService,
			folderCache: make(map[string]string),
		}, nil
	}

	// OAuth authenticated client
	if _, err := creds.GetAccessToken(ctx); err != nil {
		return nil, fmt.Errorf("failed to get access token for Drive: %w", err)
	}

	service, err := drive.NewService(ctx, option.WithTokenSource(&tokenSource{
		creds: creds,
		ctx:   ctx,
	}))
	if err != nil {
		return nil, fmt.Errorf("failed to create drive service with OAuth: %w", err)
	}

	docsService, err := docs.NewService(ctx, option.WithTokenSource(&tokenSource{
		creds: creds,
		ctx:   ctx,
	}))
	if err != nil {
		return nil, fmt.Errorf("failed to create docs service with OAuth: %w", err)
	}

	return &DriveClient{
		service:     service,
		docsService: docsService,
		oauthCreds:  creds,
		folderCache: make(map[string]string),
	}, nil
}

// tokenSource は oauth2.TokenSource インターフェースを実装
type tokenSource struct {
	creds *OAuthCredentials
	ctx   context.Context
}

func (ts *tokenSource) Token() (*oauth2.Token, error) {
	accessToken, err := ts.creds.GetAccessToken(ts.ctx)
	if err != nil {
		return nil, err
	}
	return &oauth2.Token{
		AccessToken: accessToken,
	}, nil
}

// GetFile はファイル情報を取得
func (c *DriveClient) GetFile(ctx context.Context, fileID string) (*model.FileInfo, error) {
	file, err := c.service.Files.Get(fileID).
		Fields("id, name, mimeType, parents").
		SupportsAllDrives(true).
		Context(ctx).
		Do()
	if err != nil {
		return nil, fmt.Errorf("failed to get file: %w", err)
	}

	return &model.FileInfo{
		ID:       file.Id,
		Name:     file.Name,
		MimeType: file.MimeType,
		Parents:  file.Parents,
	}, nil
}

// DownloadFile はファイルをダウンロード（堅牢化版）
func (c *DriveClient) DownloadFile(ctx context.Context, fileID string) ([]byte, error) {
	maxRetries := 5
	for attempt := 0; attempt < maxRetries; attempt++ {
		log.Printf("ダウンロード開始（試行 %d/%d）: %s", attempt+1, maxRetries, fileID)

		// リトライ時はサービスを再構築（トークン切れなどの対策）
		if attempt > 0 {
			log.Println("サービスオブジェクトを再構築してリトライします")
			var service *drive.Service
			var err error
			if c.oauthCreds != nil {
				service, err = drive.NewService(ctx, option.WithTokenSource(&tokenSource{
					creds: c.oauthCreds,
					ctx:   ctx,
				}))
			} else {
				service, err = drive.NewService(ctx, option.WithScopes(drive.DriveScope))
			}
			if err != nil {
				log.Printf("サービス再構築エラー: %v", err)
				time.Sleep(time.Duration(1<<uint(attempt)) * time.Second)
				continue
			}
			c.service = service
		}

		resp, err := c.service.Files.Get(fileID).Download()
		if err != nil {
			if attempt == maxRetries-1 {
				return nil, fmt.Errorf("download failed after %d attempts: %w", maxRetries, err)
			}
			log.Printf("ダウンロード失敗（リトライ待ち）: %v", err)
			time.Sleep(time.Duration(1<<uint(attempt)) * time.Second)
			continue
		}
		defer resp.Body.Close()

		data, err := io.ReadAll(resp.Body)
		if err != nil {
			if attempt == maxRetries-1 {
				return nil, fmt.Errorf("failed to read response body: %w", err)
			}
			log.Printf("レスポンス読み取り失敗（リトライ待ち）: %v", err)
			time.Sleep(time.Duration(1<<uint(attempt)) * time.Second)
			continue
		}

		log.Printf("ダウンロード完了: %d bytes", len(data))
		return data, nil
	}

	return nil, fmt.Errorf("download failed after %d attempts", maxRetries)
}

// MoveFile はファイルを移動
func (c *DriveClient) MoveFile(ctx context.Context, fileID string, newParentID string) error {
	// 現在の親フォルダIDを取得
	file, err := c.GetFile(ctx, fileID)
	if err != nil {
		return fmt.Errorf("failed to get file: %w", err)
	}

	if len(file.Parents) == 0 {
		return fmt.Errorf("file has no parents")
	}

	currentParentID := file.Parents[0]

	// ファイルを移動
	_, err = c.service.Files.Update(fileID, nil).
		AddParents(newParentID).
		RemoveParents(currentParentID).
		Fields("id, parents").
		SupportsAllDrives(true).
		Context(ctx).
		Do()
	if err != nil {
		return fmt.Errorf("failed to move file: %w", err)
	}

	log.Printf("ファイル移動成功: %s", fileID)
	return nil
}

// RenameFile はファイル名を変更
func (c *DriveClient) RenameFile(ctx context.Context, fileID string, newName string) error {
	_, err := c.service.Files.Update(fileID, &drive.File{
		Name: newName,
	}).SupportsAllDrives(true).Context(ctx).Do()
	if err != nil {
		return fmt.Errorf("failed to rename file: %w", err)
	}

	log.Printf("ファイル名変更成功: %s", newName)
	return nil
}

// GetOrCreateFolder はフォルダを取得または作成（排他制御付き）
func (c *DriveClient) GetOrCreateFolder(ctx context.Context, folderName string, parentID string) (string, error) {
	cacheKey := fmt.Sprintf("%s:%s", parentID, folderName)

	// 排他制御で競合を防止
	c.folderMu.Lock()
	defer c.folderMu.Unlock()

	// キャッシュを確認
	if cachedID, exists := c.folderCache[cacheKey]; exists {
		return cachedID, nil
	}

	// フォルダが存在するか検索
	query := fmt.Sprintf("name='%s' and '%s' in parents and mimeType='application/vnd.google-apps.folder' and trashed=false", folderName, parentID)
	fileList, err := c.service.Files.List().
		Q(query).
		Fields("files(id, name)").
		Context(ctx).
		Do()
	if err != nil {
		return "", fmt.Errorf("failed to search folder: %w", err)
	}

	// フォルダが存在する場合はキャッシュに追加して返す
	if len(fileList.Files) > 0 {
		folderID := fileList.Files[0].Id
		c.folderCache[cacheKey] = folderID
		return folderID, nil
	}

	// フォルダを新規作成
	folder := &drive.File{
		Name:     folderName,
		MimeType: "application/vnd.google-apps.folder",
		Parents:  []string{parentID},
	}

	createdFolder, err := c.service.Files.Create(folder).
		Fields("id").
		Context(ctx).
		Do()
	if err != nil {
		return "", fmt.Errorf("failed to create folder: %w", err)
	}

	// キャッシュに追加
	c.folderCache[cacheKey] = createdFolder.Id

	log.Printf("フォルダ作成成功: %s (%s)", folderName, createdFolder.Id)
	return createdFolder.Id, nil
}

// ListFilesInFolder はフォルダ内のファイル一覧を取得
func (c *DriveClient) ListFilesInFolder(ctx context.Context, folderID string, limit int) ([]*model.FileInfo, error) {
	query := fmt.Sprintf("'%s' in parents and trashed=false", folderID)
	call := c.service.Files.List().
		Q(query).
		PageSize(int64(limit)).
		Fields("files(id, name, mimeType, parents)").
		SupportsAllDrives(true).
		IncludeItemsFromAllDrives(true).
		Context(ctx)

	fileList, err := call.Do()
	if err != nil {
		return nil, fmt.Errorf("failed to list files: %w", err)
	}

	var files []*model.FileInfo
	for _, f := range fileList.Files {
		files = append(files, &model.FileInfo{
			ID:       f.Id,
			Name:     f.Name,
			MimeType: f.MimeType,
			Parents:  f.Parents,
		})
	}

	return files, nil
}

// GetAbout はストレージ情報を取得
func (c *DriveClient) GetAbout(ctx context.Context) (map[string]interface{}, error) {
	about, err := c.service.About.Get().
		Fields("storageQuota, user").
		Context(ctx).
		Do()
	if err != nil {
		return nil, fmt.Errorf("failed to get about info: %w", err)
	}

	result := map[string]interface{}{
		"storageQuota": about.StorageQuota,
		"user":         about.User,
	}

	return result, nil
}

// CleanupServiceAccountStorage はSAのストレージをクリーンアップ
func (c *DriveClient) CleanupServiceAccountStorage(ctx context.Context) (map[string]interface{}, error) {
	// クリーンアップロジックを実装（必要に応じて）
	// 現時点では簡易的な実装
	stats := map[string]interface{}{
		"cleaned": 0,
		"message": "Cleanup completed",
	}

	return stats, nil
}

// WatchInfo はWatch登録情報を保持
type WatchInfo struct {
	ChannelID      string
	ResourceID     string
	Expiration     int64
	StartPageToken string
}

// StartWatch はフォルダの変更監視を開始
func (c *DriveClient) StartWatch(ctx context.Context, webhookURL string) (*WatchInfo, error) {
	// 変更開始トークンを取得
	startPageToken, err := c.service.Changes.GetStartPageToken().Context(ctx).Do()
	if err != nil {
		return nil, fmt.Errorf("failed to get start page token: %w", err)
	}

	// チャンネルIDを生成（ユニークな識別子）
	channelID := fmt.Sprintf("homedocmanager-%d", time.Now().UnixNano())

	// 有効期限を7日後に設定（Drive APIの最大値）
	expiration := time.Now().Add(7 * 24 * time.Hour).UnixMilli()

	// Watch登録
	channel := &drive.Channel{
		Id:         channelID,
		Type:       "web_hook",
		Address:    webhookURL,
		Expiration: expiration,
	}

	watchedChannel, err := c.service.Changes.Watch(startPageToken.StartPageToken, channel).Context(ctx).Do()
	if err != nil {
		return nil, fmt.Errorf("failed to start watch: %w", err)
	}

	log.Printf("Watch started: channelID=%s, resourceID=%s, expiration=%d",
		watchedChannel.Id, watchedChannel.ResourceId, watchedChannel.Expiration)

	return &WatchInfo{
		ChannelID:      watchedChannel.Id,
		ResourceID:     watchedChannel.ResourceId,
		Expiration:     watchedChannel.Expiration,
		StartPageToken: startPageToken.StartPageToken,
	}, nil
}

// StopWatch は変更監視を停止
func (c *DriveClient) StopWatch(ctx context.Context, channelID, resourceID string) error {
	channel := &drive.Channel{
		Id:         channelID,
		ResourceId: resourceID,
	}

	err := c.service.Channels.Stop(channel).Context(ctx).Do()
	if err != nil {
		return fmt.Errorf("failed to stop watch: %w", err)
	}

	log.Printf("Watch stopped: channelID=%s", channelID)
	return nil
}

// GetChanges は変更されたファイルIDを取得
func (c *DriveClient) GetChanges(ctx context.Context, pageToken string) ([]string, string, error) {
	var fileIDs []string
	nextPageToken := pageToken

	for {
		changes, err := c.service.Changes.List(nextPageToken).
			Fields("nextPageToken, newStartPageToken, changes(fileId, file(id, name, mimeType, parents, trashed))").
			Context(ctx).
			Do()
		if err != nil {
			return nil, pageToken, fmt.Errorf("failed to get changes: %w", err)
		}

		for _, change := range changes.Changes {
			// 削除されたファイルや、ファイル情報がないものはスキップ
			if change.File == nil || change.File.Trashed {
				continue
			}

			// PDFまたは画像のみ対象
			mimeType := change.File.MimeType
			if mimeType == "application/pdf" ||
				mimeType == "image/jpeg" ||
				mimeType == "image/png" ||
				mimeType == "image/gif" {
				fileIDs = append(fileIDs, change.FileId)
				log.Printf("Change detected: %s (%s)", change.File.Name, change.FileId)
			}
		}

		if changes.NewStartPageToken != "" {
			nextPageToken = changes.NewStartPageToken
			break
		}
		if changes.NextPageToken != "" {
			nextPageToken = changes.NextPageToken
		} else {
			break
		}
	}

	return fileIDs, nextPageToken, nil
}
